internal/provider/services/storage/storage_test.go

```diff
@@ -0,0 +1,370 @@
+package storage
+
+import (
+	"bytes"
+	"errors"
+	mocks2 "github.com/terrariumcloud/terrarium/internal/storage/mocks"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	terrarium "github.com/terrariumcloud/terrarium/internal/provider/services"
+	"github.com/terrariumcloud/terrarium/internal/provider/services/mocks"
+	"github.com/terrariumcloud/terrarium/pkg/terrarium/provider"
+	"google.golang.org/grpc"
+)
+
+type ClosingBuffer struct {
+	*bytes.Buffer
+}
+
+func (cb *ClosingBuffer) Close() error {
+	return nil
+}
+
+// Test_RegisterStorageWithServer checks:
+// - if there was no error with bucket init
+// - if error was returned when bucket init fails
+func Test_RegisterStorageWithServer(t *testing.T) {
+	t.Parallel()
+
+	t.Run("when bucket init is successful", func(t *testing.T) {
+		s3Client := &mocks2.S3{}
+
+		ss := &StorageService{Client: s3Client}
+
+		s := grpc.NewServer(*new([]grpc.ServerOption)...)
+
+		err := ss.RegisterWithServer(s)
+
+		if err != nil {
+			t.Errorf("Expected no error, got %v.", err)
+		}
+
+		if s3Client.HeadBucketInvocations != 1 {
+			t.Errorf("Expected 1 call to HeadBucket, got %v.", s3Client.HeadBucketInvocations)
+		}
+
+		if s3Client.CreateBucketInvocations != 0 {
+			t.Errorf("Expected no calls to CreateBucket, got %v.", s3Client.CreateBucketInvocations)
+		}
+	})
+
+	t.Run("when bucket init fails", func(t *testing.T) {
+		s3Client := &mocks2.S3{
+			HeadBucketError:   errors.New("some error"),
+			CreateBucketError: errors.New("some error"),
+		}
+
+		vms := &StorageService{Client: s3Client}
+
+		s := grpc.NewServer(*new([]grpc.ServerOption)...)
+
+		err := vms.RegisterWithServer(s)
+
+		if err != BucketInitializationError {
+			t.Errorf("Expected %v, got %v.", BucketInitializationError, err)
+		}
+
+		if s3Client.HeadBucketInvocations != 1 {
+			t.Errorf("Expected 1 call to DescribeTable, got %v.", s3Client.HeadBucketInvocations)
+		}
+
+		if s3Client.CreateBucketInvocations != 1 {
+			t.Errorf("Expected 1 calls to CreateTable, got %v.", s3Client.CreateBucketInvocations)
+		}
+	})
+}
+
+// Test_DownloadProviderSourceZip checks:
+// - if correct response is returned when source zip is downloaded
+// - if error is returned when GetObject fails
+// - if error is returned when Send fails
+func Test_DownloadProviderSourceZip(t *testing.T) {
+	t.Parallel()
+
+	t.Run("when source zip is downloaded", func(t *testing.T) {
+		var length int64 = 50000
```

Comment:
This is still smaller than the default chunk size. I suggest changing the `ChunkSize` variable in the test so you don't need to have a huge buffer.
For example:
```suggestion
                chuckSize = 1000
		var length int64 = 5555
```
Note that I've made length not divisible by the chunk size to test the eof error handling.

Reply:
hmm, although tested with chunkSize = 1000 and file length to be 5555, and found the chunking algorithm to be working.
![image](https://github.com/terrariumcloud/terrarium/assets/55885026/b4ce9880-3974-44fc-a781-69dc681256f1)
we are getting the expected 6 SendInvocations & every chunk is getting read properly.
but as the ChunkSize defined will get used in the real scenario and not just limited with test, so keeping it back to 64KB and increaing the file length to be a larger value i.e. 70000, so that it tests the algorithm as well.